### PR TITLE
Application Api Keys Management

### PR DESCRIPTION
**Pull request by bafonins**
**Summary:**

This PR adds the application access add page, improves the existing ones related to api keys management and removes the api keys api stubs.

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/142

edit:
screen shot 2019-03-04 at 15 40 27

list:
screen shot 2019-02-20 at 09 52 39

add:
screen shot 2019-03-04 at 15 39 55

`ApiKeyModal`:
screen shot 2019-03-04 at 15 40 18

**Changes:**

- Add the application access add page to add new application api keys
- Improve the application access edit page
- Add the `FieldGroup` component to display errors related to a set of form fields
- Add the application api keys methods to the js sdk
- Add the application rights method to the js sdk
- Add support for injected values to the `IntlHelmet` component
- Add the `ApiKeyModal` container component to show the secret key on creation

**Notes for Reviewers:**

There are a couple of things missing:
- ~~The application rights are still taken from the fake data store. Currently, all possible rights are returned by the stack if the user is admin. It doesnt seem to be a proper behaviour and should be discussed.~~
- The rights checkbox group is not formatted as we discussed (this includes the 'check all' option, scroll). Imo, this seems like a separate PR and requires some style changes to the `Field` component.

### CODE REVIEW

File: pkg/webui/console/views/application-access-edit/index.js [hunk 1]
```diff
@@ -119,23 +173,24 @@ export default class ApplicationAccessEdit extends React.Component {
...
-      description: apiKey.description,
-      ...rightsValues,
+      rights: { ...rightsValues },
     }
 
     return (
       <Container>
         <Row>
           <Col lg={8} md={12}>
+            <IntlHelmet title={m.accessEdit} values={{ keyId }} />
```
Comment (kschiffer) [inline pkg/webui/console/views/application-access-edit/index.js]:
```suggestion
            <IntlHelmet title={m.accessEdit} values={{ keyName || 'Access Key' }} />
```
Let's not use the key id in the title, it's just an unhelpful, lengthy salad of letters. I'd suggest to use the key name and falling back to a generic 'Access Key'.

File: pkg/webui/console/views/application-access-edit/index.js [hunk 2]
```diff
@@ -18,49 +18,69 @@ import bind from 'autobind-decorator'
...
+import { getApplicationApiKey } from '../../store/actions/application'
+import api from '../../api'
 
 import style from './application-access-edit.styl'
 
 const m = defineMessages({
   deleteKey: 'Delete Key',
   modalWarning:
-    'Are you sure you want to delete "{keyName}"? Deleting an application access key cannot be undone!',
+    'Are you sure you want to delete "{keyName}"? Deleting an application access apiKey cannot be undone!',
```
[the remark this answers is not in the document]
Comment (htdvisser) [reply pkg/webui/console/views/application-access-edit/index.js]: I think we also don't say "application access API Key" in v3, just "application API key" (with lowercase key)

[the remark this answers is not in the document]
Comment (htdvisser) [reply pkg/webui/console/views/application-access-edit/index.js]: Also keep in mind that keys don't always have a name

**Comment by bafonins** [comment]
>  The Key ID should better be displayed in a safe inspector, to allow quicker interaction

The `SafeInspector` component doesnt have appropriate styling (disabled/read-only) for that. Also, the identifier is shown for completeness of the api key, not for copying.

File: pkg/webui/console/views/application-access-add/index.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      <Container>
+        <ApiKeyModal
+          {...modalProps}
+          visible={!!modal}
+          approval={false}
+        />
+        <Row>
+          <Col lg={8} md={12}>
+            <IntlHelmet title={m.accessAdd} />
+            <Message component="h2" content={sharedMessages.add} />
```
Comment (kschiffer) [inline pkg/webui/console/views/application-access-add/index.js]:
```suggestion
            <Message component="h2" content={m.accessAdd} />
```

File: pkg/webui/console/views/application-access-edit/index.js [hunk 4]
```diff
@@ -73,30 +95,68 @@ const validationSchema = Yup.object().shape({
...
+  async handleDelete () {
+    const { dispatch, appId, keyId } = this.props
+
+    await this.setState({ error: '' })
 
-    if (fetching ) {
-      return <Spinner />
+    try {
+      await api.application.apiKeys.delete(appId, keyId)
+      dispatch(replace(`/console/applications/${appId}/access`))
```
Comment (kschiffer) [inline pkg/webui/console/views/application-access-edit/index.js]: Can we show a confirmation toast here?

**Comment by kschiffer** [comment]
Also, this PR became quite big. I would prefer if we'd split the go, console and sdk bits next time, if applicable.

File: pkg/webui/console/containers/api-key-modal/index.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  return (
+    <PortalledModal
+      visible={visible}
+      modal={{
+        ...rest,
+        title: m.title,
+        subtitle: m.subtitle,
+        approval: false,
```
[the remark this answers is not in the document]
Comment (bafonins) [reply pkg/webui/console/containers/api-key-modal/index.js]: We can disable the button until the user copies the key, but I guess it is for the future

[the remark this answers is not in the document]
Comment (kschiffer) [reply pkg/webui/console/containers/api-key-modal/index.js]: I think that would cross a level of over-obtrusiveness. The user should have the freedom not to copy the key if he is informed enough to decide so.

File: pkg/webui/console/views/application-access-edit/index.js (hunk 4 above)
Comment (bafonins) [reply pkg/webui/console/views/application-access-edit/index.js]: of what type? success/info/... ?

Comment (kschiffer) [reply pkg/webui/console/views/application-access-edit/index.js]: Success.

File: pkg/webui/console/views/application-access-edit/index.js [hunk 6]
```diff
@@ -18,49 +18,72 @@ import bind from 'autobind-decorator'
...
   modalWarning:
-    'Are you sure you want to delete "{keyName}"? Deleting an application access key cannot be undone!',
+    'Are you sure you want to delete the {keyName} API Key? Deleting an application API Key cannot be undone!',
+  accessEdit: 'Edit {keyName}',
+  updateSuccess: 'Successfully updated API Key',
+  deleteSuccess: 'Successfully deleted API Key',
 })
 
 const validationSchema = Yup.object().shape({
   name: Yup.string()
```
Comment (kschiffer) [inline pkg/webui/console/views/application-access-edit/index.js]: There is a mismatch here. In the edit view, the name is required, whereas in the add new view, it is not required. The name should not be required regardless.

**Pull request merged**